libs/prelude/Prelude/Interfaces.idr

```diff
@@ -267,77 +267,96 @@ interface Foldable t where
   foldlM fm a0 = foldl (\ma, b => ma >>= flip fm b) (pure a0)
 
   ||| Produce a list of the elements contained in the parametrised type.
   toList : t elem -> List elem
   toList = foldr (::) []
 
-||| Maps each element to a value and combine them
-public export
-foldMap : (Foldable t, Monoid m) => (a -> m) -> t a -> m
-foldMap f = foldr ((<+>) . f) neutral
+  ||| Maps each element to a value and combine them.
+  ||| For performance reasons, this should wherever
+  ||| be implemented with tail recursion.
+  ||| @ f The function to apply to each element.
+  foldMap : Monoid m => (f : a -> m) -> t a -> m
+  foldMap f = foldr ((<+>) . f) neutral
 
 ||| Combine each element of a structure into a monoid.
 public export
 concat : (Foldable t, Monoid a) => t a -> a
-concat = foldr (<+>) neutral
+concat = foldMap id
 
 ||| Combine into a monoid the collective results of applying a function to each
 ||| element of a structure.
 public export
 concatMap : (Foldable t, Monoid m) => (a -> m) -> t a -> m
-concatMap f = foldr ((<+>) . f) neutral
+concatMap = foldMap
 
 ||| The conjunction of all elements of a structure containing lazy boolean
 ||| values.  `and` short-circuits from left to right, evaluating until either an
 ||| element is `False` or no elements remain.
 public export
 and : Foldable t => t (Lazy Bool) -> Bool
-and = foldl (&&) True
+and = force . concat @{(%search, monoid)}
+  where
+    monoid : Monoid (Lazy Bool)
+    monoid = MkMonoid {ty = Lazy Bool} @{MkSemigroup (\x, y => delay $ force x && y)} True
 
 ||| The disjunction of all elements of a structure containing lazy boolean
 ||| values.  `or` short-circuits from left to right, evaluating either until an
 ||| element is `True` or no elements remain.
 public export
 or : Foldable t => t (Lazy Bool) -> Bool
-or = foldl (||) False
+or = force . concat @{(%search, monoid)}
+  where
+    monoid : Monoid (Lazy Bool)
+    monoid = MkMonoid {ty = Lazy Bool} @{MkSemigroup (\x, y => x || y)} False
 
 ||| The disjunction of the collective results of applying a predicate to all
 ||| elements of a structure.  `any` short-circuits from left to right.
 public export
 any : Foldable t => (a -> Bool) -> t a -> Bool
-any p = foldl (\x,y => x || p y) False
+any = foldMap @{%search} @{monoid}
+  where
+    monoid : Monoid Bool
+    monoid = MkMonoid @{MkSemigroup (\x, y => x || y)} False
 
 ||| The disjunction of the collective results of applying a predicate to all
 ||| elements of a structure.  `all` short-circuits from left to right.
 public export
 all : Foldable t => (a -> Bool) -> t a -> Bool
-all p = foldl (\x,y => x && p y)  True
+all = foldMap @{%search} @{monoid}
+  where
+    monoid : Monoid Bool
+    monoid = MkMonoid @{MkSemigroup (\x, y => x && y)} True
 
 ||| Add together all the elements of a structure.
 public export
 sum : (Foldable t, Num a) => t a -> a
-sum = foldr (+) 0
+sum = concat @{(%search, monoid)}
+  where
+    monoid : Monoid a
+    monoid = MkMonoid @{MkSemigroup (+)} 0
```

Comment:
Why not to have a top-level named implementations e.g. `[Additive] Num a => Monoid a`, like it's done in `Data.Nat` module in `base`?

Reply:
My inital thinking was to not introduce more top levels definitions than necessary, but I guess that is only important if I export them. Each of the interfaces is only used once, so it won't make a difference to code size.
I'm happy to use top-level named instances, so I'll do that when I'm next programming.